### PR TITLE
Add Insights stopping via SpatialCommand

### PR DESCRIPTION
**Pull request by m-samiec**
#### Description
https://github.com/improbableio/UnrealEngine/pull/980
https://github.com/improbableio/UnrealEngine/pull/989

Added two methods of stopping an active Insights trace and modified existing StartInsights functionality

New SpatialExecServerCmd command - StopInsights
Usage: 
SpatialExecServerCmd <server> StopInsights

Added new parameter to StartInsights command and modified existing parameters - `-tracetime`
This sets the time in seconds that a trace will be active for before stopping insights tracing.

Usage: 
SpatialExecServerCmd <server> StartInsights [-file -channel -tracetime]

Only one trace output source per session is possible currently within Unreal, so once a trace has begun (to either file or socket), that method will be used for any subsequent output. If a `-file` is set during a StartInsights call a warning is logged and the parameter is ignored.

#### Release note
TODO

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 1]
```diff
@@ -590,3 +634,17 @@ void USpatialMetrics::RemoveCustomMetric(const FString& Metric)
 		UserSuppliedMetrics.Remove(Metric);
 	}
 }
+
+bool USpatialMetrics::StartInsightsCapture(const FString& Args)
+{
+	GCycleStatsShouldEmitNamedEvents++;
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:640]: Is this what fixes the file format problems?

**Review by improbable-danny — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 2]
```diff
@@ -20,12 +23,14 @@ namespace SpatialMetricsPrivate
...
 {
 	StartInsights,
+	StopInsights,
 
 	ServerCommandsCount,
 	ServerCommandInvalid = ServerCommandsCount,
 };
 
-const FString ServerCommandNames[static_cast<uint8>(EServerCommands::ServerCommandsCount) + 1] = { TEXT("StartInsights"), TEXT("Invalid") };
+const FString ServerCommandNames[static_cast<uint8>(EServerCommands::ServerCommandsCount) + 1] = { TEXT("StartInsights"),
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:32]: If you use an UENUM you get all this and Count for free.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 3]
```diff
@@ -450,14 +455,53 @@ void USpatialMetrics::SpatialExecServerCmd(const FString& ServerName, const FStr
...
+#elif !UE_TRACE_ENABLED
+				UE_LOG(LogSpatialMetrics, Warning,
+					   TEXT("SpatialExecServerCmd: Failed to execute server StartInsights command. UE_TRACE_ENABLE not defined."));
 #else
+				if (StartInsightsCapture(Args))
+				{
+					FString TraceTimeString;
+					if (FParse::Value(*Args, TEXT("-tracetime="), TraceTimeString))
+					{
+						int32 TraceTime = FCString::Atoi(*TraceTimeString);
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:471]: If you have an empty ' ' in here you will get 0 back, is that what you want?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp (hunk 2 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:32]: UENUM gives you enum >> string, but not string >> enum. Unless I'm mistaken?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp (hunk 3 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:471]: Oh interesting. I'm not sure. What would you expect if you put `-tracetime=` ?

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:471]: either a crash or ensure with a sensible default.

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:471]: Basically some way of tracing back this error to the place it actually happened.

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:471]: I think its reasonable to assume that an empty string means a dev or tool person forgot to fill in the string.

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:471]: If this is the first in a string of arguments you might just eat a spacing space as well or consume end padding which sometimes (accidentally) exists when constructing commandline arguments.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp (hunk 2 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:32]: https://docs.unrealengine.com/4.26/en-US/API/Runtime/CoreUObject/UObject/UEnum/GetIndexByName/

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp (hunk 3 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:471]: (I have recently seen this happen in someone else's code, so its not theoretical)

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp (hunk 2 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:32]: Hmm maybe trickier than I thought.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp (hunk 3 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:471]: Okay I'll check if the value is <= 0, and warn and ignore if so

**Review by simonsarginson — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 4]
```diff
@@ -381,14 +353,125 @@ void USpatialMetrics::OnModifySettingCommand(Schema_Object* CommandPayload)
...
+}
+
+void USpatialMetrics::OnExecServerCmdCommand(Schema_Object* CommandPayload)
+{
+	const FString ServerName =
+		SpatialGDK::GetStringFromSchema(CommandPayload, SpatialConstants::EXEC_SERVER_COMMAND_PAYLOAD_SERVER_NAME_ID);
+	const int32 Command = Schema_GetInt32(CommandPayload, SpatialConstants::EXEC_SERVER_COMMAND_PAYLOAD_COMMAND_ID);
+	const FString Args = SpatialGDK::GetStringFromSchema(CommandPayload, SpatialConstants::EXEC_SERVER_COMMAND_PAYLOAD_ARGS_ID);
+
+	if (!StaticEnum<ESpatialServerCommands>()->IsValidEnumValue(Command))
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:374]: Ah nice, more UENUM goodies!

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 5]
```diff
@@ -18,39 +18,6 @@
 
 DEFINE_LOG_CATEGORY(LogSpatialMetrics);
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:20]: Nice!

**Review by simonsarginson — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 6]
```diff
@@ -449,15 +533,59 @@ void USpatialMetrics::SpatialExecServerCmd(const FString& ServerName, const FStr
...
 			switch (ServerCommand)
 			{
-			case SpatialMetricsPrivate::EServerCommands::StartInsights:
-#if ENGINE_MINOR_VERSION >= 26
-				FTraceAuxiliary::UpdateTraceCapture(*Args);
+			case ESpatialServerCommands::StartInsights:
+			{
+#if ENGINE_MINOR_VERSION < 26
+				UE_LOG(LogSpatialMetrics, Warning,
+					   TEXT("SpatialExecServerCmd: Failed to execute server StartInsights command. Command only available prior to 4.26."));
```
Comment (cmsmithio) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:540]: post-4.26?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 7]
```diff
@@ -381,14 +353,125 @@ void USpatialMetrics::OnModifySettingCommand(Schema_Object* CommandPayload)
 
 void USpatialMetrics::SpatialExecServerCmd(const FString& ServerName, const FString& Command, const FString& Args)
 {
-	const SpatialMetricsPrivate::EServerCommands ServerCommand = SpatialMetricsPrivate::ServerCommandsStringToEnum(Command);
-	if (ServerCommand == SpatialMetricsPrivate::EServerCommands::ServerCommandInvalid)
+	int32 Index = StaticEnum<ESpatialServerCommands>()->GetIndexByNameString(Command);
```
Comment (cmsmithio) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:356]: nit: const

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 8]
```diff
@@ -449,15 +533,59 @@ void USpatialMetrics::SpatialExecServerCmd(const FString& ServerName, const FStr
...
+#elif !UE_TRACE_ENABLED
+				UE_LOG(LogSpatialMetrics, Warning,
+					   TEXT("SpatialExecServerCmd: Failed to execute server StartInsights command. UE_TRACE_ENABLE not defined."));
 #else
+				if (StartInsightsCapture(Args))
+				{
+					FString TraceTimeString;
+					if (FParse::Value(*Args, TEXT("-tracetime="), TraceTimeString))
+					{
+						int32 TraceTime = FCString::Atoi(*TraceTimeString);
```
Comment (cmsmithio) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:550]: nit const

**Review by cmsmithio — APPROVED** [review]

**Review by cmsmithio — APPROVED** [review]

**Pull request merged**